src/core/unit.c

```diff
@@ -1742,6 +1745,23 @@ int unit_load(Unit *u) {
                 unit_invalidate_cgroup_members_masks(u);
         }
 
+        /* For top-level cgroups, if not set during loading, restrict allowed CPUs based on manager value */
+        if(u->manager->defaults.allowed_cpus.allocated > 0 && UNIT_HAS_CGROUP_CONTEXT(u)) {
+                cc = unit_get_cgroup_context(u);
+                if(!cc)
+                        goto fail;
+
+                if(cc->cpuset_cpus.allocated == 0) {
+                        r = unit_default_cgroup_path(u, &unit_cgroup_path);
+                        if (r < 0 || unit_cgroup_path == NULL)
+                                goto fail;
+
+                        /* All top-level cgroups should have only one leading slash */
+                        if(strchr(&unit_cgroup_path[1], '/') == NULL)
+                                cpu_set_copy(&cc->cpuset_cpus, &u->manager->defaults.allowed_cpus);
+                }
```

Comment:
please do not parse cgroup paths. this is simply wrong, because the root of the cgroup tree is not necessarily the root of the tree that systemd manages.

just do a check UNIT_GET_SLICE() on the unit to get the slice it is in. It will return NULL if you are in "-.slice".

Reply:
That makes more sense, but I'm not sure how to move it there. The parent slice property/dependency isn't set [until slice_load()](https://github.com/systemd/systemd/blob/main/src/core/slice.c#L179) (in` unit_load()`). So in `unit_init()`, I don't know how to check for top-level units.